### DISCUSSION TITLE
`SelectingItemsControl` functional way of keeping selection

### DISCUSSION THREAD

**Discussion by JaggerJo** [AvaloniaUI/Avalonia#14177] [Ideas]
It would be nice if `SelectedItemsControl` derived classes would have a way of keeping the selection even if the data collection completely changes.

Assume the data we are binding is constantly updated. Item's are not just removed & added - they also change.

```csharp
public record NetworkDevice
{
    public required string MacAdress { get; init; }
    public required DateTime LastSeenAt { get; init; } // <- constantly changes
}
```

It would be nice if the selected item(s) selection state could be determined by a custom identity accessor `Func<NetworkDevice, string>`

**Comment by maxkatz6** [comment]
This is possible on view model level, and somewhat trivial with DynamicData (though, DynamicData has high entry level anyway).

In general, you should be able to redefine Equals+GetHashCode to make it work without any extra code, can't you?

**Reply by JaggerJo** [reply]
> This is possible on view model level, and somewhat trivial with DynamicData (though, DynamicData has high entry level anyway).

@maxkatz6 I know about and have used dynamic data, but wouldn't I have the same issue with the code below if the selected `Trade` class is modified and has value equality schematics?

```csharp
ReadOnlyObservableCollection<TradeProxy> list;

var myTradeCache = new SourceCache<Trade, long>(trade => trade.Id);
var myOperation = myTradeCache.Connect() 
		.Filter(trade=>trade.Status == TradeStatus.Live) 
		.Transform(trade => new TradeProxy(trade))
		.Sort(SortExpressionComparer<TradeProxy>.Descending(t => t.Timestamp))
		.ObserveOnDispatcher()
		.Bind(out list) 
		.DisposeMany()
		.Subscribe()
```
(taken from t[he docs](https://github.com/reactivemarbles/DynamicData))

> In general, you should be able to redefine Equals+GetHashCode to make it work without any extra code, can't you?

Yes, I think this would work - but I'd rather not change the types as they are immutable and used in other places where value equality is the expected behaviour. 

The simple solution would be to have a `Device` type that only contains the non changing fields I guess. 

I just wanted to bring this up here as I'm dealing with this a lot in different places. Mostly because I build UI's on immutable data structures that are streamed via observables.

**Reply by timunie** [reply]
If you don't want to change the type, wrap it in a container ItemViewModel imo

**Reply by JaggerJo** [reply]
Yeah, that would be a solution. I currently map to a type that only contains the stable fields. 

My point still stands I think that this is not straight forward and there might be a better way to track the selected item(s).

**Reply by timunie** [reply]
Another option is to implement the SelectionModel, suer.
